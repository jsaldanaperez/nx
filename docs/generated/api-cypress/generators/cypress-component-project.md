---
title: '@nrwl/cypress:cypress-component-project generator'
description: 'Set up Cypress Component Test for a project'
---

# @nrwl/cypress:cypress-component-project

Set up Cypress Component Test for a project

## Usage

```bash
nx generate cypress-component-project ...
```

```bash
nx g cy-cmp ... # same
```

By default, Nx will search for `cypress-component-project` in the default collection provisioned in `workspace.json`.

You can specify the collection explicitly as follows:

```bash
nx g @nrwl/cypress:cypress-component-project ...
```

Show what will be generated without writing to disk:

```bash
nx g cypress-component-project ... --dry-run
```

### Examples

Add cypress component testing to an existing project named my-cool-lib:

```bash
nx g @nrwl/cypress:cypress-component-project --project=my-cool-lib --componentType=react
```

Add Cypress component testing to an existing project that already has a Cypress component testing enabled. Helpful to reset the configuration to the default.:

```bash
nx g @nrwl/cypress:cypress-component-project --project=my-cool-lib --componentType=react --force
```

## Options

### componentType (_**required**_)

Alias(es): framework

Type: `string`

Possible values: `react`, `next`

Which framework is being used?

### project (_**required**_)

Type: `string`

The name of the project to convert

### compiler

Default: `babel`

Type: `string`

Possible values: `swc`, `babel`

Which compiler is being used in the existing project?

### force

Default: `false`

Type: `boolean`

Overwrite an existing cypress component test project. This is helpful if you want to reset to the default configurations.
