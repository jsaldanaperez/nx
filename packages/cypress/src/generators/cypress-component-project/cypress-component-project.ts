import { CypressComponentProjectSchema } from './schema';
import {
  addDependenciesToPackageJson,
  formatFiles,
  generateFiles,
  joinPathFragments,
  offsetFromRoot,
  ProjectConfiguration,
  readProjectConfiguration,
  Tree,
  updateJson,
  updateProjectConfiguration,
} from '@nrwl/devkit';
import { join } from 'path';
import {
  cypressReactVersion,
  cypressVersion,
  cypressWebpackVersion,
  swcCoreVersion,
  swcLoaderVersion,
  webpackHttpPluginVersion,
} from '../../utils/versions';
import { installedCypressVersion } from '../../utils/cypress-version';
import { CYPRESS_COMPONENT_TEST_TARGET_NAME } from '../../utils/project-name';

/**
 * Check the installed version of cypress.
 * @returns true if cypress should be installed or false if cypress doesn't need to be installed, i.e. already installed or higher version is installed
 * @throws Error if this installed cypress version is not supported.
 * @throws Error if the force flag is not supplied AND either cypress.config.ts or cypress component test target already exists
 */
export function checkInstalledCypress(
  tree: Tree,
  options: CypressComponentProjectSchema,
  projectConfig: ProjectConfiguration
): boolean {
  // if the project already has the component testing target or a cypress.config.ts file then we make sure the --force flag is sent. if not, we throw an error.
  if (
    projectConfig.targets?.[CYPRESS_COMPONENT_TEST_TARGET_NAME] ||
    tree.exists(joinPathFragments(projectConfig.root, 'cypress.config.ts'))
  ) {
    if (!options.force) {
      throw new Error(
        `The project already has a cypress component testing target. Please use the --force flag to overwrite the existing project.`
      );
    }
    return true;
  }

  const installedVersion = installedCypressVersion();
  // if we don't have a version then it's safe to install cypress
  if (!installedVersion) {
    return true;
  }

  if (installedVersion >= 10) {
    return false;
  }

  throw new Error('Cypress version 10 or greater is required ');
}

function updateDeps(
  tree: Tree,
  options: CypressComponentProjectSchema,
  shouldInstallCypress: boolean
) {
  const devDeps = {
    '@cypress/webpack-dev-server': cypressWebpackVersion,
    'html-webpack-plugin': webpackHttpPluginVersion,
  };

  if (shouldInstallCypress) {
    devDeps['cypress'] = cypressVersion;
  }

  if (options.componentType === 'react' || options.componentType === 'next') {
    devDeps['@cypress/react'] = cypressReactVersion;
    if (options.compiler === 'swc') {
      devDeps['@swc/core'] = swcCoreVersion;
      devDeps['swc-loader'] = swcLoaderVersion;
    }
  }
  return addDependenciesToPackageJson(tree, {}, devDeps);
}

function addFiles(
  tree: Tree,
  projectConfig: ProjectConfiguration,
  options: CypressComponentProjectSchema
) {
  generateFiles(tree, join(__dirname, 'files'), projectConfig.root, {
    ...options,
    projectRoot: projectConfig.root,
    offsetFromRoot: offsetFromRoot(projectConfig.root),
    ext: '',
  });
}

function addTargetToProject(
  projectConfig: ProjectConfiguration,
  tree: Tree,
  options: CypressComponentProjectSchema
) {
  projectConfig.targets[CYPRESS_COMPONENT_TEST_TARGET_NAME] = {
    executor: '@nrwl/cypress:cypress',
    options: {
      cypressConfig: joinPathFragments(projectConfig.root, 'cypress.config.ts'),
      testingType: 'component',
    },
  };

  updateProjectConfiguration(tree, options.project, projectConfig);
}

export async function cypressComponentProject(
  tree: Tree,
  options: CypressComponentProjectSchema
) {
  const projectConfig = readProjectConfiguration(tree, options.project);

  const shouldInstallCypress = checkInstalledCypress(
    tree,
    options,
    projectConfig
  );

  addFiles(tree, projectConfig, options);

  updateTSConfig(tree, projectConfig);

  addTargetToProject(projectConfig, tree, options);

  const installDeps = updateDeps(tree, options, shouldInstallCypress);

  return () => {
    formatFiles(tree);
    installDeps();
  };
}

// TODO(caleb):is this actually needed?
//  it _seems_ to be fine without a tsconfig??
function updateTSConfig(tree: Tree, projectConfig: ProjectConfiguration) {
  const projectTsConfigPath = joinPathFragments(
    projectConfig.root,
    'tsconfig.json'
  );
  if (!tree.exists(projectTsConfigPath)) {
    throw new Error(
      `Expected project tsconfig.json to exist. Please create one. Expected ${projectTsConfigPath} to exist. Found none.`
    );
  }
  updateJson(tree, projectTsConfigPath, (json) => {
    json.references = json.references || [];
    json.references.push({ path: './tsconfig.cy.json' });
    return json;
  });
}
