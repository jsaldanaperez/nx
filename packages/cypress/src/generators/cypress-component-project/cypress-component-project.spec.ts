import {
  addProjectConfiguration,
  ProjectConfiguration,
  readProjectConfiguration,
  Tree,
  updateProjectConfiguration,
} from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { installedCypressVersion } from '../../utils/cypress-version';

import {
  CYPRESS_COMPONENT_TEST_TARGET_NAME,
  cypressComponentProject,
} from '@nrwl/cypress';
import { checkInstalledCypress } from './cypress-component-project';

jest.mock('../../utils/cypress-version');
let projectConfig: ProjectConfiguration = {
  projectType: 'library',
  sourceRoot: 'libs/cool-lib/src',
  root: 'libs/cool-lib',
  targets: {
    test: {
      executor: '@nrwl/jest:jest',
      options: {
        jestConfig: 'libs/cool-lib/jest.config.js',
      },
    },
  },
};
describe('Cypress Component Project', () => {
  let tree: Tree;
  let mockedInstalledCypressVersion: jest.Mock<
    ReturnType<typeof installedCypressVersion>
  > = installedCypressVersion as never;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
    addProjectConfiguration(tree, 'cool-lib', projectConfig);
    tree.write(
      'libs/cool-lib/tsconfig.json',
      `
{
  "references": [
    {
      "path": "./tsconfig.lib.json"
    },
    {
      "path": "./tsconfig.spec.json"
    }
  ]
}
`
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('validation', () => {
    it('should be able to create project with cypress 10 installed', () => {
      mockedInstalledCypressVersion.mockReturnValue(10);
      const actual = checkInstalledCypress(
        tree,
        {
          project: 'cool-lib',
          compiler: 'babel',
          componentType: 'react',
          force: false,
        },
        projectConfig
      );
      expect(actual).toEqual(false);
    });

    it('should not be able to create project with < cypress v9 installed', () => {
      mockedInstalledCypressVersion.mockReturnValue(9);

      expect(() =>
        checkInstalledCypress(
          tree,
          {
            project: 'cool-lib',
            compiler: 'swc',
            componentType: 'react',
            force: false,
          },
          projectConfig
        )
      ).toThrowErrorMatchingSnapshot();
    });

    it('should throw an error if cypress project is already created', () => {
      mockedInstalledCypressVersion.mockReturnValue(9);
      tree.write('libs/cool-lib/cypress.config.ts', '');
      expect(() =>
        checkInstalledCypress(
          tree,
          {
            project: 'cool-lib',
            compiler: 'swc',
            componentType: 'react',
            force: false,
          },
          projectConfig
        )
      ).toThrowErrorMatchingSnapshot();
    });

    it('should create cypress project over existing with --force', () => {
      mockedInstalledCypressVersion.mockReturnValue(9);
      tree.write('libs/cool-lib/cypress.config.ts', '');
      const newTarget = {
        [CYPRESS_COMPONENT_TEST_TARGET_NAME]: {
          executor: '@nrwl/cypress:cypress',
          options: {
            cypressConfig: 'libs/cool-lib/cypress.config.ts',
            testingType: 'component',
          },
        },
      };
      updateProjectConfiguration(tree, 'cool-lib', {
        ...projectConfig,
        targets: {
          ...projectConfig.targets,
          ...newTarget,
        },
      });

      const actual = checkInstalledCypress(
        tree,
        {
          project: 'cool-lib',
          compiler: 'swc',
          componentType: 'react',
          force: true,
        },
        projectConfig
      );
      expect(actual).toEqual(true);
    });
  });

  describe('react', () => {
    beforeEach(() => {
      mockedInstalledCypressVersion.mockReturnValue(10);
    });
    it('should create project w/babel', async () => {
      tree.delete('libs/cool-lib/cypress.config.ts');
      await cypressComponentProject(tree, {
        project: 'cool-lib',
        compiler: 'babel',
        componentType: 'react',
        force: false,
      });
      const actualProjectConfig = readProjectConfiguration(tree, 'cool-lib');

      expect(tree.exists('libs/cool-lib/cypress.config.ts')).toEqual(true);
      expect(tree.exists('libs/cool-lib/cypress')).toEqual(true);
      expect(tree.exists('libs/cool-lib/tsconfig.cy.json')).toEqual(true);
      expect(actualProjectConfig.targets['test-cmp']).toMatchSnapshot();
    });

    it('should create project w/swc', async () => {
      tree.delete('libs/cool-lib/cypress.config.ts');
      await cypressComponentProject(tree, {
        project: 'cool-lib',
        compiler: 'swc',
        componentType: 'react',
        force: false,
      });

      const actualProjectConfig = readProjectConfiguration(tree, 'cool-lib');

      expect(tree.exists('libs/cool-lib/cypress.config.ts')).toEqual(true);
      expect(tree.exists('libs/cool-lib/cypress')).toEqual(true);
      expect(tree.exists('libs/cool-lib/tsconfig.cy.json')).toEqual(true);
      expect(actualProjectConfig.targets['test-cmp']).toMatchSnapshot();
    });
  });
});
