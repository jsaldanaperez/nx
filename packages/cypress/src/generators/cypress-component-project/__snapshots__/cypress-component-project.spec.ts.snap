// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cypress Component Project react should create project w/babel 1`] = `
Object {
  "executor": "@nrwl/cypress:cypress",
  "options": Object {
    "cypressConfig": "libs/cool-lib/cypress.config.ts",
    "testingType": "component",
  },
}
`;

exports[`Cypress Component Project react should create project w/swc 1`] = `
Object {
  "executor": "@nrwl/cypress:cypress",
  "options": Object {
    "cypressConfig": "libs/cool-lib/cypress.config.ts",
    "testingType": "component",
  },
}
`;

exports[`Cypress Component Project validation should not be able to create project with < cypress v9 installed 1`] = `"Cypress version 10 or greater is required "`;

exports[`Cypress Component Project validation should throw an error if cypress project is already created 1`] = `"The project already has a cypress component testing target. Please use the --force flag to overwrite the existing project."`;
